src/coreclr/tools/dotnet-pgo/TypeRefTypeSystem/TypeRefTypeSystemGenericParameter.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Internal.TypeSystem;
+
+namespace Microsoft.Diagnostics.Tools.Pgo.TypeRefTypeSystem
+{
+    public sealed class TypeRefTypeSystemGenericParameter : GenericParameterDesc
+    {
+        private TypeRefTypeSystemType _owningType;
+        private TypeRefTypeSystemMethod _owningMethod;
+        private int _index;
+
+        internal TypeRefTypeSystemGenericParameter(TypeRefTypeSystemType owningType, int index)
+        {
+            _owningType = owningType;
+            _index = index;
+        }
+
+        internal TypeRefTypeSystemGenericParameter(TypeRefTypeSystemMethod owningMethod, int index)
+        {
+            _owningMethod = owningMethod;
+            _index = index;
+        }
+
+        public override TypeSystemContext Context
+        {
+            get
+            {
+                return _owningType.Context;
+            }
+        }
+
+        public override string Name
+        {
+            get
+            {
+                return $"_{_index}";
```

Comment:
Nit - wouldn't it be more expressive to use '!' to refer to generic parameters akin to the standard MSIL notation?

Reply:
Yes. Done.